### PR TITLE
feat: add `filter_by` method to the `ArgillaTrainer` 

### LINKED ISSUE #3646: [FEATURE] add `filter_by` method to the `ArgillaTrainer`
**Is your feature request related to a problem? Please describe.**
From this PR, it is possible to filter the a dataset.

**Describe the solution you'd like**
I want to use this to filter on submitted records by default and allow for changing this via a filter_by arg during trainer definition and the `train` method.

**Describe alternatives you've considered**
N.A.

**Additional context**
#3610

### PR DESCRIPTION
**Pull request by plaguss**
# Description

This PR adds the functionality to filter by response_status on the `ArgillaTrainer`:

```python
from argilla import SortBy

trainer = ArgillaTrainer(
    dataset=remote_feedback_dataset,
    task=task,
    framework=framework,
    filter_by={"response_status": ["submitted"]},
    sort_by=[SortBy(field="metadata.my-metadata", order="asc")]
)
```

Closes #3646 and #4096

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor (change restructuring the codebase without changing functionality)
- [x] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

- [x] `tests/integration/client/feedback/test_trainer.py::test_trainer_with_filter_by_and_sort_by`

**Checklist**

- [x] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [x] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla/client/feedback/training/base.py [hunk 1]
```diff
@@ -71,10 +71,20 @@ def __init__(
...
                 will be used by default. GPU IDs start in 0, which stands for the default GPU in the system,
                 if available.
+            filter_by: A dict with key the field to filter by, and values the filters to apply. Currently only
+                defined for `response_status` filters. Can be one of: draft, pending, submitted, and discarded.
+                Defaults to `None` (no filter is applied).
             framework_kwargs: arguments for the framework's trainer. A special key (model_card_kwargs) is reserved
                 for the arguments that can be passed to the model card.
             **load_kwargs: arguments for the rg.load() function.
         """
+        if filter_by:
```
Comment (davidberenstein1957) [inline src/argilla/client/feedback/training/base.py:85]: I think we always need to pull RemoteFeedbackDatasets. Also could you include the sort_by as well?

File: src/argilla/client/feedback/training/base.py [hunk 2]
```diff
@@ -71,10 +71,20 @@ def __init__(
...
+            # Use the id attribute of RemoteFeedbackDataset to determine the type of dataset.
+            # There are errors due to circular imports when trying to import RemoteFeedbackDataset.
+            if not getattr(dataset, "id", None):
```
Comment (davidberenstein1957) [inline src/argilla/client/feedback/training/base.py]: I think a `isinstance` check for `RemoteFeedbackDataset` might be better.

**Review by davidberenstein1957 — CHANGES_REQUESTED** [review]
Some minor remarks. Perhaps we can add the sort_by and could you mention it in the docs too?

File: src/argilla/client/feedback/training/base.py (hunk 2 above)
Comment (plaguss) [reply src/argilla/client/feedback/training/base.py]: I also think that's the best option, but I kept errors regarding circular options. Maybe we can review this problem tomorrow?

File: src/argilla/client/feedback/training/base.py (hunk 1 above)
Comment (plaguss) [reply src/argilla/client/feedback/training/base.py:85]: Included in [0b36807](https://github.com/argilla-io/argilla/pull/4120/commits/0b3680797b98d28e499551ec3b79073bfc365ffc)

File: src/argilla/client/feedback/training/base.py (hunk 2 above)
Comment (plaguss) [reply src/argilla/client/feedback/training/base.py]: There isn't really a need for that check after all, it's removed now.

**Comment by plaguss** [comment]
Let's add the docs

File: docs/_source/practical_guides/fine_tune.md [hunk 3]
```diff
@@ -82,6 +82,26 @@ A `TrainingTask` is used to define how the data should be processed and formatte
...
+    dataset=dataset,
+    task=task,
+    framework="setfit",
+    filter_by={"response_status": ["submitted"]},
+    sort_by=[SortBy(field="metadata.my-metadata", order="asc")],
+)
+```
+
+:::{note}
+You can take a look at [filter and query datasets](./filter_dataset.md) page in the docs to learn more about how to filter and sort datasets.
```
Comment (davidberenstein1957) [inline docs/_source/practical_guides/fine_tune.md]: can you make this a relative path from roots?
```suggestion
You can take a look at [filter and query datasets](/practical_guides/fine_tune.md) page in the docs to learn more about how to filter and sort datasets.
```

**Review by davidberenstein1957 — CHANGES_REQUESTED** [review]

**Pull request merged**